### PR TITLE
added file logging of observations

### PR DESCRIPTION
**Pull request by vincentberenz**
Implementation of a python API for dumping the observations generated by a robot backend into files.

This is documented here:
https://github.com/intelligent-soft-robots/intelligent-soft-robots.github.io/wiki/Usage_01_Saving-data-in-file

This works as follow:
a process is spawned. The process encapsulate a robot frontend. The robot frontend then calls at a desired frequency the method get_observations_since, which returns all observations that have not been save in the file yet. The observations are serialized (using cereal as exposed by the shared_memory package) and dumped into the file.

For this to work, an instance of the c++ shared_memory::Serializer had to also be exposed by the python API

### CODE REVIEW

File: python/o80_pam/logger.py [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+        self.num = num # num counter
+        self.date = date # date
+
+    @staticmethod
+    def create(prefix,path):
+        # given an path (i.e. /tmp/o80_pam_observations_num_date)
+        # and a prefix (o80_pam_observations)
+        # creates a corresponding instance of _File
+        # (i.e. extracts num and date from the path)
+        filename = path[path.find(prefix):]
```
Comment (luator) [inline python/o80_pam/logger.py:161]: There might be some special cases, I'm missing but wouldn't do `os.path.basename(path)` do the same and be easier to understand?

File: python/o80_pam/logger.py [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+    
+    def __init__(self,path,num,date):
+        self.path = path # absolute path to the file (name included)
+        self.num = num # num counter
+        self.date = date # date
+
+    @staticmethod
+    def create(prefix,path):
+        # given an path (i.e. /tmp/o80_pam_observations_num_date)
+        # and a prefix (o80_pam_observations)
```
Comment (luator) [inline python/o80_pam/logger.py:158]:
```suggestion
        # and a prefix (o80_pam_observations_)
```

File: bin/o80_log_file_stats [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+        # saving info about the first Observations
+        if first_iteration is None:
+            first_iteration = iteration
+            first_time_stamp = time_stamp
+            current_iteration = iteration
+            first_observation = observation
+        # if 2 observation's iteration are separate by more than 1,
+        # it means the logger missed one observation during recording
+        if iteration-current_iteration>1:
+            missed_iterations += 1
```
Comment (luator) [inline bin/o80_log_file_stats:34]: I think if more than one observation is missing in a row, this would only count one of them.

**Review by luator — COMMENTED** [review]
Looks pretty nice :)  I have a few comments below but nothing critical.

One suggestion regarding the log file format:  It might be useful in the future to add a small header with a bit of meta data.  In robot_interfaces I'm simply adding a version number for the log file format (see [here](https://github.com/open-dynamic-robot-initiative/robot_interfaces/blob/6ebc4c043b1b0739814bd278d253b195a3e05b2a/include/robot_interfaces/robot_logger.hpp#L272)), which I increase, if there is some change that is not compatible with already existing log files.  So when reading the file I can detect if the file has an older format and print a more useful error message (or could potentially have different loaders for different formats).
In the case here, it might be useful to add the expected `serialized_size` for sanity checking.

**Comment by vincentberenz** [comment]
@luator I made the recommended fixes, and also added unit tests (better late than never). I think it's ready to merge, but please let me know if you'd prefer to review the updates

File: bin/o80_log_file_stats [hunk 4]
```diff
@@ -31,7 +31,7 @@ def _print_stats(file_path):
         # if 2 observation's iteration are separate by more than 1,
         # it means the logger missed one observation during recording
         if iteration-current_iteration>1:
-            missed_iterations += 1
+            missed_iterations += (iteration-current_iteration)
```
Comment (luator) [inline bin/o80_log_file_stats:34]: I think it should be `interation - current_iteration - 1` as only the iterations up to `iteration` have been missed but not `iteration` itself.

File: python/o80_pam/logger.py [hunk 5]
```diff
@@ -73,6 +73,14 @@ class Logger:
     
     def __init__(self,segment_id,file_path,frequency=500.):
 
+        # throwing exception if the folder of file_path
+        # does not exists or is not writable
+        filename = os.path.basename(file_path)
+        folder = file_path[:-len(filename)]
```
Comment (luator) [inline python/o80_pam/logger.py:79]: `folder = os.path.dirname(file_path)` should do the same.

File: tests/test_logger.py [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+class O80_PAM_LOGGER_TESTCASE(unittest.TestCase):
+
+    def test_exception_on_invalid_path(self):
+        try:
+            o80_pam.Logger("foo_id",os.path.join("not","existing"))
+            raised=False
+        except FileNotFoundError:
+            raised=True
+        # note: self.assertRaises did not work
+        self.assertTrue(raised)
```
Comment (luator) [inline tests/test_logger.py:17]: unittest has an `assertRaises` which can be used to simplify this:
```python
with self.assertRaises(FileNotFoundError):
    o80_pam.Logger("foo_id",os.path.join("not","existing"))
```

File: tests/test_logger.py [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+                      "get_velocities",
+                      "get_positions"):
+                self.assertEqual(getattr(o1,f)(),
+                                 getattr(o2,f)())
+            self.assertEqual(o1.get_desired_states().get(dof).get(),
+                             o2.get_desired_states().get(dof).get())
+            self.assertEqual(o1.get_observed_states().get(dof).get(),
+                             o2.get_observed_states().get(dof).get())
+
+    def test_logger_no_directory(self):
```
Comment (luator) [inline tests/test_logger.py:39]: Isn't this a duplicate of `test_exception_on_invalid_path`?

**Review by luator — APPROVED** [review]
If I didn't confuse things, the `missed_iterations` calculation is off by one.  Apart from this, it looks good :+1:

**Comment by vincentberenz** [comment]
@luator actually good that you looked in details :) Will do the fixes

File: tests/test_logger.py (hunk 6 above)
Comment (vincentberenz) [reply tests/test_logger.py:17]: That was my first version, for some unknown reason, it did not work (i.e. the test would not pass)

**Pull request merged**